lib/github.com/diku-dk/containers/setops.fut

```diff
@@ -0,0 +1,141 @@
+import "../sorts/radix_sort"
+import "opt"
+
+-- | Set-like operations on arrays of values interpreted as sets
+
+local
+-- | The setops module type.  This module type is declared `local`, which means
+-- that it may not be referenced directly by name from client code.  This
+-- limitation makes it possible for the interface to be enriched by new members
+-- in future minor versions.
+
+module type setops = {
+  -- | [union a b] returns the set of elements that appear in both a and b with
+  -- dublicates removed. Work: O(n+m), span: O(1).
+  val union               [m][n] : [m]i64 -> [n]i64 -> ?[k].[k]i64
+
+  -- | [intersect a b] returns those elements in a that also appears in b with
+  -- dublicates removed. Work: O(n+m), span: O(1).
+  val intersect           [m][n] : [m]i64 -> [n]i64 -> ?[k].[k]i64
+
+  -- | [diff a b] returns a with dublicates and elements from b removed. Work:
+  -- O(n+m), span: O(1).
+  val diff                [m][n] : [m]i64 -> [n]i64 -> ?[k].[k]i64
+
+  -- | [elimdubs a] returns a with dublicates removed. Work: O(n), span: O(1).
+  val elimdubs               [m] : [m]i64 -> ?[k].[k]i64
+
+  -- | [union_by_key key f a b] returns the set of elements that appear in both
+  -- a and b, identified using the function key, with dublicates removed, and
+  -- with the function f applied to elements that occur in both a and b. Work:
+  -- O(n+m), span: O(1), assuming key and f have work complexity O(1).
+  val union_by_key     't [m][n] : (t->i64) -> (t->t->t) -> [m]t -> [n]t -> ?[k].[k]t
+
+  -- | [intersect_by_key key f a b] returns those elements in a that also
+  -- appears in b, identified using the function key, with dublicates removed,
+  -- and with the function f applied to the intersecting objects. Work: O(n+m),
+  -- span: O(1), assuming key and f have work complexity O(1).
+  val intersect_by_key 't [m][n] : (t->i64) -> (t->t->t) -> [m]t -> [n]t -> ?[k].[k]t
+
+  -- | [diff_by_key key a b] returns a with dublicates and elements from b
+  -- removed, where the function key is used for identifying objects. Work:
+  -- O(n+m), span: O(1), assuming key has work complexity O(1).
+  val diff_by_key      't [m][n] : (t->i64) -> [m]t -> [n]t -> ?[k].[k]t
+
+  -- | [join_by_key key1 key2 a b] returns pairs of objects in a and b that
+  -- agrees on keys obtained with the key1 and key2 functions.
+  val join_by_key   'a 'b [m][n] : (a->i64) -> (b->i64) -> [m]a -> [n]b -> ?[k].[k](a,b)
+
+  -- | [elimdubs_by_key key f a] returns a with dublicates removed, identified
+  -- using the function key, and with dublicates merged using the function f,
+  -- which is assumed to be associative. Work: O(n), span: O(1), assuming key
+  -- and f have work complexity O(1).
+  val elimdubs_by_key     't [m] : (t->i64) -> (t->t->t) -> [m]t -> ?[k].[k]t
+}
+
+
+module setops : setops = {
+
+  def unsome 'a f (x:opt a) : a =
```

Comment:
We have this as `from_opt` (although with a slightly different type).

Reply:
`from_opt` doesn't work for me...